nav2_waypoint_follower/src/waypoint_follower.cpp

```diff
@@ -304,4 +329,174 @@ WaypointFollower::goalResponseCallback(
   }
 }
 
+void WaypointFollower::followGPSWaypoints()
+{
+  auto goal = gps_action_server_->get_current_goal();
+  auto feedback = std::make_shared<ActionTGPS::Feedback>();
+  auto result = std::make_shared<ActionTGPS::Result>();
+
+  if (!gps_action_server_ || !gps_action_server_->is_server_active()) {
+    RCLCPP_DEBUG(get_logger(), "GPS Action server inactive. Stopping.");
+    return;
+  }
+
+  RCLCPP_INFO(
+    get_logger(), "Received follow GPS waypoint request with %i waypoints.",
+    static_cast<int>(goal->waypoints.size()));
+
+  std::vector<geometry_msgs::msg::PoseStamped> poses = convertGPSWaypointstoPosesinMap(
+    goal->waypoints, shared_from_this(), from_ll_to_map_client_);
+
+  auto is_action_server_ready =
+    waypoint_follower_action_client_->wait_for_action_server(std::chrono::seconds(5));
+  if (!is_action_server_ready) {
+    RCLCPP_ERROR(
+      this->get_logger(), "FollowWaypoints action server is not available."
+      "make an instance of waypoint_follower is up and running");
+    return;
+  }
+  waypoint_follower_goal_ = ClientTGPS::Goal();
+  waypoint_follower_goal_.poses = poses;
+
+  RCLCPP_INFO(
+    this->get_logger(),
+    "Sending a path of %zu waypoints:", waypoint_follower_goal_.poses.size());
+  for (auto waypoint : waypoint_follower_goal_.poses) {
+    RCLCPP_DEBUG(
+      this->get_logger(),
+      "\t(%lf, %lf)", waypoint.pose.position.x, waypoint.pose.position.y);
+  }
+
+  auto goal_options = ActionClientGPS::SendGoalOptions();
+  goal_options.result_callback = std::bind(
+    &WaypointFollower::GPSResultCallback, this,
+    std::placeholders::_1);
+  goal_options.goal_response_callback = std::bind(
+    &WaypointFollower::GPSGoalResponseCallback, this,
+    std::placeholders::_1);
+
+  auto future_goal_handle = waypoint_follower_action_client_->async_send_goal(
+    waypoint_follower_goal_, goal_options);
+  if (rclcpp::spin_until_future_complete(
+      waypoint_follower_client_node_,
+      future_goal_handle) != rclcpp::FutureReturnCode::SUCCESS)
+  {
+    RCLCPP_ERROR(get_logger(), "Send goal call failed");
+    return;
+  }
+  // Get the goal handle and save so that we can check
+  // on completion in the timer callback
+  waypoint_follower_goal_handle_ = future_goal_handle.get();
+  if (!waypoint_follower_goal_handle_) {
+    RCLCPP_ERROR(get_logger(), "Goal was rejected by server");
+    return;
+  }
+}
+
+std::vector<geometry_msgs::msg::PoseStamped>
+WaypointFollower::convertGPSWaypointstoPosesinMap(
+  const
+  std::vector<sensor_msgs::msg::NavSatFix> & gps_waypoints,
+  const rclcpp_lifecycle::LifecycleNode::SharedPtr & parent_node,
+  const rclcpp::Client<robot_localization::srv::FromLL>::SharedPtr & fromll_client)
+{
+  RCLCPP_INFO(parent_node->get_logger(), "Converting GPS waypoints to Map Frame..");
+  if (!parent_node) {
```

Comment:
This is promised to exist. `shared_from_this` I believe will throw its own exception if it isn't able to. I don't think the checks on parent_node how you're using it is necessary.